### DISCUSSION TITLE
A few questions regarding CLI usage with a free account

### DISCUSSION THREAD

**Discussion by ZaLiTHkA** [tolgee/tolgee-platform#2420] [Q&A]
howdy good people,

I'm relatively new to Tolgee, found it little over a week back and I've successfully used the web interface to translate all of the strings in my hybrid mobile app. this morning I tried to see if I could integrate the CLI tool with my workspace, as this seems to be the most efficient way to handle synchronising the list of "translation keys" in my local project with the associated Tolgee project. but this is where I'm starting to run into some issues..

for context, my workspace is a Nx monorepo with multiple Angular-based and Capacitor-powered hybrid mobile applications.

the application I'm working with currently has 360 keys and 2 languages, which does still fit within the 1000 string limit of the free tier. so I created a config file at `./apps/my-app/.tolgeerc.json` (file extension simply to ensure my IDE uses the appropriate syntax highlighting), and set my `push` and `pull` paths appropriately (`./src/assets/i18n/` relative to my `.tolgeerc.json` file, because that's where Transloco looks for the files).

---

first point: I then tried to run `tolgee -c ./apps/my-app/.tolgeerc.json push`, since my local translation files did have a few changes (4 keys added, 2 keys removed), but at the "importing" step I got the message `Invalid request data [status: 400, code: plan_translation_limit_exceeded]`.

that message makes me think the CLI tool is checking how many strings I have available to me, seeing that it's less than it's about to try uploading, and then not going any further. or am I misunderstanding the meaning of this error code?

---

second point: I then tried to run `tolgee -c ./apps/my-app/.tolgeerc.json compare`, hoping that might give a little more context as to what the Tolgee CLI tool was seeing from the local and remote sources, but this gave me the message `No files were matched for extraction`.

did I configure my project incorrectly for this? I'm not trying to use any of the automated extraction features, and I do not intend to. so I'm not sure what I'm missing here...

---

third point: I then made a backup of my local files and ran `tolgee -c ./apps/my-app/.tolgeerc.json pull`, which successfully downloaded my `en.json` and `de.json` files from Tolgee to the correct place locally, but using the traditional "nested JSON" structure, not the "Flat JSON" structure that I get when exporting through the web UI.

I do a string search in my workspace when looking for translation keys, which helps to keep consistency between my translation and markup files. so can the CLI tool (currently on `2.1.5`) not export translations as "flat JSON"? is this something that is going to be added at some point?

---

note: if you need any more details regarding my workspace configuration, please don't hesitate to ask. if I can get this platform to fit in my workflow, even if that means adjusting my workflow, there is a very good chance I can convince my employer to sign us up for a paid subscription, where we will be managing a few thousand translation keys across multiple hybrid-mobile, desktop, and web interfaces.. just need to work out a few teething issues first.

**Comment by stepan662** [comment]
Hello, hopefully I can answer all your points:

1. The error message is coming from backend, and it indicates that the import would exceed the 1000 strings limit (which is weird since you shouldn't have more then 1000). You can try to import each language separately and see if it increases number of strings, or use the manual import in the UI, that might be a bit more user friendly to debug this.

2. The `compare` command extracts keys from your code and compares them to tolgee platform (it's not working with your json translations), so you can see, if you have any unused ones etc. We don't yet have extractor for angular, so it will not work for you currently (although we want to fix this in near future).

3. I think you can set `format: JSON_ICU`, for structured json. https://tolgee.io/tolgee-cli/project-configuration#format

**Reply by ZaLiTHkA** [reply]
thanks @stepan662, that is helpful indeed.. done a bit more testing now, but unfortunately I do still have a few blocking issues.

> 1. The error message is coming from backend, and it indicates that the import would exceed the 1000 strings limit (which is weird since you shouldn't have more then 1000). You can try to import each language separately and see if it increases number of strings, or use the manual import in the UI, that might be a bit more user friendly to debug this.

this is essentially what I thought was happening, though I don't think this is something I can fix from project configuration... I would say the CLI tool needs to check the key names and calculate the final number of keys before making this decision.

from a pure numbers point of view, importing my current `360` strings, even for just one language, _should_ exceed my free account's `1000` limit if they were **new** keys.. because I already have this project configured with 2 languages, putting me on `718` strings (2 of my entries have not been translated yet). but these are not new, I get the exact same behaviour when I try push a file that I just pulled from my Tolgee project without making any changes.

the main reason I was trying to use the CLI for this, is because I needed to removed some translation keys, and change some others. so I understood from the docs that the CLI would allow me to "sync" my local and remote keys, removing old keys from the remote project when I remove them locally.

is it possible to remove "missing keys" when importing files through the web UI? effectively replicating the `--remove-other-keys` flag for the CLI `push` command?

> 2. The `compare` command extracts keys from your code and compares them to tolgee platform (it's not working with your json translations), so you can see, if you have any unused ones etc. We don't yet have extractor for angular, so it will not work for you currently (although we want to fix this in near future).

that makes a bit more sense now, and based on that, it's something I highly doubt I would ever use.

> 3. I think you can set `format: JSON_ICU`, for structured json. [tolgee.io/tolgee-cli/project-configuration#format](https://tolgee.io/tolgee-cli/project-configuration#format)

just tried that and it made no difference at all, exporting with both the `JSON_ICU` and `JSON_TOLGEE` formats present absolutely no differences in my project. I did try a few other formats, just to make sure my configuration file was actually being used, and I did indeed get my translations in the file format I specified.

I suspect that my translation strings simply don't contain anything that would change how Tolgee handles "ICU placeholders", though I'm not intimately familiar with them, so that's just speculation on my part.

**Reply by ZaLiTHkA** [reply]
just found something in [an issue](https://github.com/tolgee/tolgee-cli/issues/55) logged in the `tolgee-cli` project that helped a bit.. adding `"delimiter": null` to the `pull` object in my Tolgee config file gave me the output I was looking for.

still having trouble with pushing local changes to my hosted Tolgee project though. should I perhaps rather be using `sync` for that then? (edit) or maybe not, that seems to be part of the "string extraction" logic, which is not what I'm looking for.

**Reply by stepan662** [reply]
Ah, yeah, you are right about the delimiter. No the `sync` command also works with keys extracted from code, so that is something else.

The tolgee CLI doesn't really check number of strings in your project, it simply tries to import the files to the platform and the platform rejects it, if it goes above your limit.

So what I was suggesting was to try to import them manually throught the platform UI (which is essentially the same thing) and see if there is some problem (like it's trying to create new keys instead of updating existing ones).

**Reply by ZaLiTHkA** [reply]
right... I think I understand, conflict resolution is all done server-side, hence the existence of the `--force-mode` CLI flag when pushing translations.

just discovered something interesting though: if I run `tolgee push` from my Windows 11 Pro environment, I get the aforementioned `plan_translation_limit_exceeded` error response, but the same command with the same files from my Zorin OS (Debian based) environment works correctly.

I'll do a little more testing with verbose output and log an issue in the CLI project shortly.

**Reply by ZaLiTHkA** [reply]
now I just need you (@stepan662) to reply with a RTFM comment so I can mark that as the answer to my questions... 🤣

**Reply by stepan662** [reply]
Hmm, yeah, if you'd be able to investigate a bit more, that would be helpful. No, your questions are to the point, no worries 😄

**Reply by ZaLiTHkA** [reply]
will do, thank you for the assistance. :)
